sway-lib-std/src/u128.sw

```diff
@@ -0,0 +1,310 @@
+    }
+
+    pub fn lt(self, other: Self) -> bool {
+        self.upper < other.upper || self.upper == other.upper && self.lower < other.lower
+    }
+}
+
+// TODO this doesn't work?
+// impl core::ops::OrdEq for U128 {
+// }
+
+fn disable_overflow() {
+    asm(r1) {
+        movi r1 i3;
+        flag r1;
+    }
+}
+
+fn enable_overflow() {
+    asm(r1) {
+        movi r1 i0;
+        flag r1;
+    }
+}
+
+impl u64 {
+    pub fn overflowing_add(a: u64, b: u64) -> U128 {
+        disable_overflow();
+        let mut v = U128 {
+            upper: 0,
+            lower: 0,
+        };
+        asm(r1, r2, a: a, b: b, v_ptr: v) {
+            add r1 a b;
+            move r2 of;
+            sw v_ptr r2 i0;
+            sw v_ptr r1 i1;
+        };
+        enable_overflow();
+        v
+    }
+
+    pub fn overflowing_mul(self, b: Self) -> U128 {
+        disable_overflow();
+        let mut v = U128 {
+            upper: 0,
+            lower: 0,
+        };
+        asm(r1, r2, a: self, b: b, v_ptr: v) {
+            mul r1 a b;
+            move r2 of;
+            sw v_ptr r2 i0;
+            sw v_ptr r1 i1;
+        };
+        enable_overflow();
+        v
+    }
+}
+
+impl U128 {
+    /// Initializes a new, zeroed U128.
+    pub fn new() -> U128 {
+        U128 {
+            upper: 0,
+            lower: 0,
+        }
+    }
+
+    /// Downcast to `u64`. Err if precision would be lost, Ok otherwise.
+    pub fn to_u64(self) -> Result<u64, ()> {
+        match self.upper {
+            0 => {
+                Result::Ok(self.lower)
+            },
+            _ => {
+                Result::Err(())
+            },
+        }
+    }
+
+    /// The smallest value that can be represented by this integer type.
+    pub fn min() -> U128 {
+        U128 {
+            upper: 0,
+            lower: 0,
+        }
+    }
+
+    /// The largest value that can be represented by this type,
+    /// 2<sup>128</sup> - 1.
+    pub fn max() -> U128 {
+        U128 {
+            upper: ~u64::max(),
+            lower: ~u64::max(),
+        }
+    }
+
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        128
+    }
+}
+
+impl core::ops::BitwiseAnd for U128 {
+    pub fn binary_and(self, other: Self) -> Self {
+        ~U128::from(self.upper & other.upper, self.lower & other.lower)
+    }
+}
+
+impl core::ops::BitwiseOr for U128 {
+    pub fn binary_or(self, other: Self) -> Self {
+        ~U128::from(self.upper | other.upper, self.lower | other.lower)
+    }
+}
+
+impl core::ops::Shiftable for U128 {
+    pub fn lsh(self, other: u64) -> Self {
+        // If shifting by at least the number of bits, then saturate with
+        // zeroes.
+        if (other >= 128) {
+            return ~Self::new();
+        };
+
+        // If shifting by at least half the number of bits, then upper word can
+        // be discarded.
+        if (other >= 64) {
+            return ~Self::from(self.lower << (other - 64), 0);
```

Comment:
The compiler emits an incorrect warning that this line is unreachable. cc @otrho

Reply:
#1682